Remove() causes panic: runtime error: index out of range

This:

```
cache = ttlcache.NewCache()
cache.Set("key", "value")
//cache.Remove("key")
count = cache.Count()
fmt.Printf("cache has %d keys\n", count)
```

prints this:

> cache has 1 keys

But if you uncomment the Remove() call, you get this:

> panic: runtime error: index out of range
> goroutine 1 [running]:
> panic(0x76ed00, 0xc820014120)
>     /software/vertres/installs/go/src/runtime/panic.go:464 +0x3e6
> github.com/diegobernardes/ttlcache.priorityQueue.Swap(0x0, 0x0, 0x0, 0x0, 0x0, 0xffffffffffffffff)
>     /nfs/users/nfs_s/sb10/src/go/src/github.com/diegobernardes/ttlcache/priority_queue.go:54 +0x1f9
> github.com/diegobernardes/ttlcache.(_priorityQueue).Swap(0xc8200d8020, 0x0, 0xffffffffffffffff)
>     <autogenerated>:3 +0xaa
> container/heap.Remove(0x7f595f9ac750, 0xc8200d8020, 0x0, 0x0, 0x0)
>     /software/vertres/installs/go/src/container/heap/heap.go:74 +0x6c
> github.com/diegobernardes/ttlcache.(_priorityQueue).remove(0xc8200d8020, 0xc8200e0000)
>     /nfs/users/nfs_s/sb10/src/go/src/github.com/diegobernardes/ttlcache/priority_queue.go:35 +0x56
> github.com/diegobernardes/ttlcache.(*Cache).Remove(0xc8200de000, 0x7daeb8, 0x3, 0x6dd860)
>     /nfs/users/nfs_s/sb10/src/go/src/github.com/diegobernardes/ttlcache/cache.go:167 +0xed
> [...]

I've had this problem too.

panic: runtime error: index out of range

goroutine 12 [running]:
panic(0x826680, 0xc82000a0d0)
        C:/Go/src/runtime/panic.go:481 +0x3e6
github.com/diegobernardes/ttlcache.priorityQueue.Swap(0x0, 0x0, 0x0, 0x0, 0x0, 0xffffffffffffffff)
        d:/workspace/src/github.com/diegobernardes/ttlcache/priority_queue.go:54 +0x1f9
github.com/diegobernardes/ttlcache.(*priorityQueue).Swap(0xc82000e2e0, 0x0, 0xffffffffffffffff)
        <autogenerated>:3 +0xaa
container/heap.Remove(0x7f74086d5118, 0xc82000e2e0, 0x0, 0x0, 0x0)
        C:/Go/src/container/heap/heap.go:74 +0x6c
github.com/diegobernardes/ttlcache.(*priorityQueue).remove(0xc82000e2e0, 0xc8203bfcc0)
        d:/workspace/src/github.com/diegobernardes/ttlcache/priority_queue.go:35 +0x56
github.com/diegobernardes/ttlcache.(*Cache).Remove(0xc8201241e0, 0xc820495330, 0x9, 0x0)
        d:/workspace/src/github.com/diegobernardes/ttlcache/cache.go:167 +0xed
main.(*Factory).Put(0xc820014380, 0x0, 0x0)
        d:/workspace/src/m-services/m-fence/factory.go:120 +0x460
main.(*Factory).DeliveLoop(0xc820014380)
        d:/workspace/src/m-services/m-fence/factory.go:100 +0x30
created by main.main
        d:/workspace/src/m-services/m-fence/main.go:34 +0x5e9

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,7 +2,8 @@ language: go
 
 go:
   - 1.4.2
-
+  - 1.7
+  - 1.8
 git:
   depth: 1
 

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,7 +1,6 @@
 language: go
 
 go:
-  - 1.4.2
   - 1.7
   - 1.8
 git:

--- a/Readme.md
+++ b/Readme.md
@@ -1,8 +1,3 @@
-# READ BEFORE USE!
-This project have some problems, don't use it in production. I'm really out of time to fix it now, so if you must use this project, please, fix it first and send a PR or join as a contributor.
-
-If you really need something ready to use just take a look [here](https://github.com/avelino/awesome-go#data-structures)
-
 ## TTLCache - an in-memory cache with expiration
 
 TTLCache is a simple key/value cache in golang with the following functions:

--- a/cache.go
+++ b/cache.go
@@ -1,6 +1,8 @@
 package ttlcache
 
 import (
+	"log"
+	"sort"
 	"sync"
 	"time"
 )
@@ -40,10 +42,10 @@ func (cache *Cache) getItem(key string) (*item, bool) {
 
 		item.touch()
 		cache.priorityQueue.update(item)
-		cache.expirationNotificationTrigger(item)
 	}
 
 	cache.mutex.Unlock()
+	cache.expirationNotificationTrigger(item)
 	return item, exists
 }
 
@@ -52,11 +54,14 @@ func (cache *Cache) startExpirationProcessing() {
 		var sleepTime time.Duration
 		cache.mutex.Lock()
 		if cache.priorityQueue.Len() > 0 {
-			if cache.ttl > 0 && time.Now().Add(cache.ttl).Before(cache.priorityQueue.items[0].expireAt) {
-				sleepTime = cache.ttl
-			} else {
-				sleepTime = cache.priorityQueue.items[0].expireAt.Sub(time.Now())
+			sleepTime = cache.priorityQueue.items[0].expireAt.Sub(time.Now())
+			if sleepTime < 0 {
+				sleepTime = time.Duration(time.Hour)
 			}
+			if cache.ttl > 0 {
+				sleepTime = min(sleepTime, cache.ttl)
+			}
+
 		} else if cache.ttl > 0 {
 			sleepTime = cache.ttl
 		} else {
@@ -65,37 +70,38 @@ func (cache *Cache) startExpirationProcessing() {
 
 		cache.expirationTime = time.Now().Add(sleepTime)
 		cache.mutex.Unlock()
-
+		log.Printf("Sleeping for: %v", sleepTime)
 		select {
-		case <-time.After(cache.expirationTime.Sub(time.Now())):
-			if cache.priorityQueue.Len() == 0 {
-				continue
-			}
-
+		case <-time.After(sleepTime):
 			cache.mutex.Lock()
-			item := cache.priorityQueue.items[0]
-
-			if !item.expired() {
+			if cache.priorityQueue.Len() == 0 {
 				cache.mutex.Unlock()
 				continue
 			}
 
-			if cache.checkExpireCallback != nil {
-				if !cache.checkExpireCallback(item.key, item.data) {
-					item.touch()
-					cache.priorityQueue.update(item)
-					cache.mutex.Unlock()
-					continue
+			for item := cache.priorityQueue.items[0]; item.expired(); item = cache.priorityQueue.items[0] {
+
+				if cache.checkExpireCallback != nil {
+					if !cache.checkExpireCallback(item.key, item.data) {
+						item.touch()
+						cache.priorityQueue.update(item)
+						continue
+					}
+				}
+
+				cache.priorityQueue.remove(item)
+				delete(cache.items, item.key)
+				if cache.expireCallback != nil {
+					go cache.expireCallback(item.key, item.data)
+				}
+				if cache.priorityQueue.Len() == 0 {
+					goto done
 				}
 			}
-
-			cache.priorityQueue.remove(item)
-			delete(cache.items, item.key)
+			done:
 			cache.mutex.Unlock()
 
-			if cache.expireCallback != nil {
-				cache.expireCallback(item.key, item.data)
-			}
+
 		case <-cache.expirationNotification:
 			continue
 		}
@@ -103,8 +109,12 @@ func (cache *Cache) startExpirationProcessing() {
 }
 
 func (cache *Cache) expirationNotificationTrigger(item *item) {
+	cache.mutex.Lock()
 	if cache.expirationTime.After(time.Now().Add(item.ttl)) {
+		cache.mutex.Unlock()
 		cache.expirationNotification <- true
+	} else {
+		cache.mutex.Unlock()
 	}
 }
 
@@ -138,12 +148,14 @@ func (cache *Cache) SetWithTTL(key string, data interface{}, ttl time.Duration) 
 	} else {
 		cache.priorityQueue.push(item)
 	}
-	cache.expirationNotificationTrigger(item)
+
+	sort.Sort(cache.priorityQueue)
 
 	cache.mutex.Unlock()
 	if !exists && cache.newItemCallback != nil {
 		cache.newItemCallback(key, data)
 	}
+	cache.expirationNotification <- true
 }
 
 // Get is a thread-safe way to lookup items
@@ -181,8 +193,8 @@ func (cache *Cache) Count() int {
 func (cache *Cache) SetTTL(ttl time.Duration) {
 	cache.mutex.Lock()
 	cache.ttl = ttl
-	cache.expirationNotification <- true
 	cache.mutex.Unlock()
+	cache.expirationNotification <- true
 }
 
 // SetExpirationCallback sets a callback that will be called when an item expires
@@ -211,4 +223,11 @@ func NewCache() *Cache {
 	}
 	go cache.startExpirationProcessing()
 	return cache
+}
+
+func min(duration time.Duration, second time.Duration) time.Duration {
+	if duration < second {
+		return duration
+	}
+	return second
 }

--- a/cache.go
+++ b/cache.go
@@ -130,16 +130,16 @@ func (cache *Cache) SetWithTTL(key string, data interface{}, ttl time.Duration) 
 		if cache.ttl > 0 && item.ttl == 0 {
 			item.ttl = cache.ttl
 		}
-
 		item.touch()
-
-		if exists {
-			cache.priorityQueue.update(item)
-		} else {
-			cache.priorityQueue.push(item)
-		}
-		cache.expirationNotificationTrigger(item)
 	}
+
+	if exists {
+		cache.priorityQueue.update(item)
+	} else {
+		cache.priorityQueue.push(item)
+	}
+	cache.expirationNotificationTrigger(item)
+
 	cache.mutex.Unlock()
 	if !exists && cache.newItemCallback != nil {
 		cache.newItemCallback(key, data)

--- a/cache.go
+++ b/cache.go
@@ -1,7 +1,6 @@
 package ttlcache
 
 import (
-	"log"
 	"sync"
 	"time"
 )

--- a/cache.go
+++ b/cache.go
@@ -2,7 +2,6 @@ package ttlcache
 
 import (
 	"log"
-	"sort"
 	"sync"
 	"time"
 )
@@ -70,7 +69,7 @@ func (cache *Cache) startExpirationProcessing() {
 
 		cache.expirationTime = time.Now().Add(sleepTime)
 		cache.mutex.Unlock()
-		log.Printf("Sleeping for: %v", sleepTime)
+
 		select {
 		case <-time.After(sleepTime):
 			cache.mutex.Lock()
@@ -79,12 +78,15 @@ func (cache *Cache) startExpirationProcessing() {
 				continue
 			}
 
-			for item := cache.priorityQueue.items[0]; item.expired(); item = cache.priorityQueue.items[0] {
+			// index will only be advanced if the current entry will not be evicted
+			i := 0
+			for item := cache.priorityQueue.items[i]; item.expired(); item = cache.priorityQueue.items[i] {
 
 				if cache.checkExpireCallback != nil {
 					if !cache.checkExpireCallback(item.key, item.data) {
 						item.touch()
 						cache.priorityQueue.update(item)
+						i++
 						continue
 					}
 				}
@@ -148,8 +150,6 @@ func (cache *Cache) SetWithTTL(key string, data interface{}, ttl time.Duration) 
 	} else {
 		cache.priorityQueue.push(item)
 	}
-
-	sort.Sort(cache.priorityQueue)
 
 	cache.mutex.Unlock()
 	if !exists && cache.newItemCallback != nil {

--- a/cache.go
+++ b/cache.go
@@ -99,9 +99,8 @@ func (cache *Cache) startExpirationProcessing() {
 					goto done
 				}
 			}
-			done:
+		done:
 			cache.mutex.Unlock()
-
 
 		case <-cache.expirationNotification:
 			continue
@@ -210,6 +209,14 @@ func (cache *Cache) SetCheckExpirationCallback(callback checkExpireCallback) {
 // SetNewItemCallback sets a callback that will be called when a new item is added to the cache
 func (cache *Cache) SetNewItemCallback(callback expireCallback) {
 	cache.newItemCallback = callback
+}
+
+// Purge will remove all entries
+func (cache *Cache) Purge() {
+	cache.mutex.Lock()
+	cache.items = make(map[string]*item)
+	cache.priorityQueue = newPriorityQueue()
+	cache.mutex.Unlock()
 }
 
 // NewCache is a helper to create instance of the Cache struct

--- a/cache_test.go
+++ b/cache_test.go
@@ -148,6 +148,8 @@ func TestCacheExpirationCallbackFunction(t *testing.T) {
 	assert.Equal(t, 2, expiredCount, "Expected 2 items to be expired")
 }
 
+// TestCacheCheckExpirationCallbackFunction should consider that the next entry in the queue
+// needs to be considered for eviction even if the callback returns no eviction for the current item
 func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 	expiredCount := 0
 	var lock sync.Mutex
@@ -155,7 +157,7 @@ func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 	cache := NewCache()
 	cache.SetTTL(time.Duration(50 * time.Millisecond))
 	cache.SetCheckExpirationCallback(func(key string, value interface{}) bool {
-		if key == "key2" {
+		if key == "key2" || key == "key4"{
 			return true
 		}
 		return false
@@ -166,7 +168,11 @@ func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 		lock.Unlock()
 	})
 	cache.Set("key", "value")
+	cache.Set("key3", "value")
 	cache.Set("key2", "value")
+	cache.Set("key4", "value")
+
+
 	<-time.After(110 * time.Millisecond)
 	lock.Lock()
 	assert.Equal(t, 1, expiredCount, "Expected 1 item to be expired")

--- a/cache_test.go
+++ b/cache_test.go
@@ -132,15 +132,15 @@ func TestCacheExpirationCallbackFunction(t *testing.T) {
 	var lock sync.Mutex
 
 	cache := NewCache()
-	cache.SetTTL(time.Duration(50 * time.Millisecond))
+	cache.SetTTL(time.Duration(500 * time.Millisecond))
 	cache.SetExpirationCallback(func(key string, value interface{}) {
 		lock.Lock()
 		defer lock.Unlock()
 		expiredCount = expiredCount + 1
 	})
-	cache.SetWithTTL("key", "value", time.Duration(100*time.Millisecond))
+	cache.SetWithTTL("key", "value", time.Duration(1000*time.Millisecond))
 	cache.Set("key_2", "value")
-	<-time.After(110 * time.Millisecond)
+	<-time.After(1100 * time.Millisecond)
 
 	lock.Lock()
 	defer lock.Unlock()

--- a/cache_test.go
+++ b/cache_test.go
@@ -3,7 +3,7 @@ package ttlcache
 import (
 	"testing"
 	"time"
-	
+
 	"github.com/stretchr/testify/assert"
 	"sync"
 )
@@ -156,7 +156,7 @@ func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 	cache := NewCache()
 	cache.SetTTL(time.Duration(50 * time.Millisecond))
 	cache.SetCheckExpirationCallback(func(key string, value interface{}) bool {
-		if key == "key2" || key == "key4"{
+		if key == "key2" || key == "key4" {
 			return true
 		}
 		return false
@@ -170,7 +170,6 @@ func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 	cache.Set("key3", "value")
 	cache.Set("key2", "value")
 	cache.Set("key4", "value")
-
 
 	<-time.After(110 * time.Millisecond)
 	lock.Lock()
@@ -213,6 +212,23 @@ func TestCacheSetWithTTLExistItem(t *testing.T) {
 	data, exists := cache.Get("key")
 	assert.Equal(t, true, exists, "Expected 'key' to exist")
 	assert.Equal(t, "value2", data.(string), "Expected 'data' to have value 'value2'")
+}
+
+func TestCache_Purge(t *testing.T) {
+	cache := NewCache()
+	cache.SetTTL(time.Duration(100 * time.Millisecond))
+
+	for i := 0; i < 5; i++ {
+
+		cache.SetWithTTL("key", "value", time.Duration(50*time.Millisecond))
+		<-time.After(30 * time.Millisecond)
+		cache.SetWithTTL("key", "value2", time.Duration(50*time.Millisecond))
+		cache.Get("key")
+
+		cache.Purge()
+		assert.Equal(t, 0, cache.Count(), "Cache should be empty")
+	}
+
 }
 
 func BenchmarkCacheSetWithoutTTL(b *testing.B) {

--- a/cache_test.go
+++ b/cache_test.go
@@ -4,8 +4,48 @@ import (
 	"testing"
 	"time"
 
+	"fmt"
 	"github.com/stretchr/testify/assert"
 )
+
+// test github issue #4
+func TestRemovalAndCountDoesNotPanic(t *testing.T) {
+	cache := NewCache()
+	cache.Set("key", "value")
+	cache.Remove("key")
+	count := cache.Count()
+	fmt.Printf("cache has %d keys\n", count)
+}
+
+// test github issue #3
+func TestRemovalWithTtlDoesNotPanic(t *testing.T) {
+	cache := NewCache()
+	cache.SetExpirationCallback(func(key string, value interface{}) {
+		fmt.Printf("This key(%s) has expired\n", key)
+	})
+
+	cache.SetWithTTL("keyWithTTL", "value", time.Duration(2*time.Second))
+	cache.Set("key", "value")
+	cache.Remove("key")
+
+	value, exists := cache.Get("keyWithTTL")
+	if exists {
+		fmt.Printf("got %s for keyWithTTL\n", value)
+	}
+	count := cache.Count()
+	fmt.Printf("cache has %d keys\n", count)
+
+	<-time.After(3 * time.Second)
+
+	value, exists = cache.Get("keyWithTTL")
+	if exists {
+		fmt.Printf("got %s for keyWithTTL\n", value)
+	} else {
+		fmt.Println("keyWithTTL has gone")
+	}
+	count = cache.Count()
+	fmt.Printf("cache has %d keys\n", count)
+}
 
 func TestCacheIndividualExpirationBiggerThanGlobal(t *testing.T) {
 	cache := NewCache()

--- a/cache_test.go
+++ b/cache_test.go
@@ -3,8 +3,7 @@ package ttlcache
 import (
 	"testing"
 	"time"
-
-	"fmt"
+	
 	"github.com/stretchr/testify/assert"
 	"sync"
 )
@@ -15,7 +14,7 @@ func TestRemovalAndCountDoesNotPanic(t *testing.T) {
 	cache.Set("key", "value")
 	cache.Remove("key")
 	count := cache.Count()
-	fmt.Printf("cache has %d keys\n", count)
+	t.Logf("cache has %d keys\n", count)
 }
 
 // test github issue #3
@@ -31,21 +30,21 @@ func TestRemovalWithTtlDoesNotPanic(t *testing.T) {
 
 	value, exists := cache.Get("keyWithTTL")
 	if exists {
-		fmt.Printf("got %s for keyWithTTL\n", value)
+		t.Logf("got %s for keyWithTTL\n", value)
 	}
 	count := cache.Count()
-	fmt.Printf("cache has %d keys\n", count)
+	t.Logf("cache has %d keys\n", count)
 
 	<-time.After(3 * time.Second)
 
 	value, exists = cache.Get("keyWithTTL")
 	if exists {
-		fmt.Printf("got %s for keyWithTTL\n", value)
+		t.Logf("got %s for keyWithTTL\n", value)
 	} else {
-		fmt.Println("keyWithTTL has gone")
+		t.Logf("keyWithTTL has gone")
 	}
 	count = cache.Count()
-	fmt.Printf("cache has %d keys\n", count)
+	t.Logf("cache has %d keys\n", count)
 }
 
 func TestCacheIndividualExpirationBiggerThanGlobal(t *testing.T) {

--- a/cache_test.go
+++ b/cache_test.go
@@ -6,6 +6,7 @@ import (
 
 	"fmt"
 	"github.com/stretchr/testify/assert"
+	"sync"
 )
 
 // test github issue #4
@@ -21,7 +22,7 @@ func TestRemovalAndCountDoesNotPanic(t *testing.T) {
 func TestRemovalWithTtlDoesNotPanic(t *testing.T) {
 	cache := NewCache()
 	cache.SetExpirationCallback(func(key string, value interface{}) {
-		fmt.Printf("This key(%s) has expired\n", key)
+		t.Logf("This key(%s) has expired\n", key)
 	})
 
 	cache.SetWithTTL("keyWithTTL", "value", time.Duration(2*time.Second))
@@ -82,17 +83,20 @@ func TestCacheGlobalExpiration(t *testing.T) {
 	cache.Set("key_1", "value")
 	cache.Set("key_2", "value")
 	<-time.After(200 * time.Millisecond)
-	assert.Equal(t, cache.Count(), 0, "Cache should be empty")
-	assert.Equal(t, cache.priorityQueue.Len(), 0, "PriorityQueue should be empty")
+	assert.Equal(t, 0, cache.Count(), "Cache should be empty")
+	assert.Equal(t, 0, cache.priorityQueue.Len(), "PriorityQueue should be empty")
 }
 
 func TestCacheMixedExpirations(t *testing.T) {
 	cache := NewCache()
+	cache.SetExpirationCallback(func(key string, value interface{}) {
+		t.Logf("expired: %s", key)
+	})
 	cache.Set("key_1", "value")
 	cache.SetTTL(time.Duration(100 * time.Millisecond))
 	cache.Set("key_2", "value")
-	<-time.After(200 * time.Millisecond)
-	assert.Equal(t, cache.Count(), 1, "Cache should have only 1 item")
+	<-time.After(150 * time.Millisecond)
+	assert.Equal(t, 1, cache.Count(), "Cache should have only 1 item")
 }
 
 func TestCacheIndividualExpiration(t *testing.T) {
@@ -108,7 +112,7 @@ func TestCacheIndividualExpiration(t *testing.T) {
 	cache.SetWithTTL("key4", "value", time.Duration(50*time.Millisecond))
 	<-time.After(100 * time.Millisecond)
 	<-time.After(100 * time.Millisecond)
-	assert.Equal(t, cache.Count(), 0, "Cache should be empty")
+	assert.Equal(t, 0, cache.Count(), "Cache should be empty")
 }
 
 func TestCacheGet(t *testing.T) {
@@ -120,25 +124,34 @@ func TestCacheGet(t *testing.T) {
 	cache.Set("hello", "world")
 	data, exists = cache.Get("hello")
 	assert.NotNil(t, data, "Expected data to be not nil")
-	assert.Equal(t, exists, true, "Expected data to exist")
-	assert.Equal(t, (data.(string)), "world", "Expected data content to be 'world'")
+	assert.Equal(t, true, exists, "Expected data to exist")
+	assert.Equal(t, "world", (data.(string)), "Expected data content to be 'world'")
 }
 
 func TestCacheExpirationCallbackFunction(t *testing.T) {
 	expiredCount := 0
+	var lock sync.Mutex
+
 	cache := NewCache()
 	cache.SetTTL(time.Duration(50 * time.Millisecond))
 	cache.SetExpirationCallback(func(key string, value interface{}) {
+		lock.Lock()
+		defer lock.Unlock()
 		expiredCount = expiredCount + 1
 	})
 	cache.SetWithTTL("key", "value", time.Duration(100*time.Millisecond))
 	cache.Set("key_2", "value")
 	<-time.After(110 * time.Millisecond)
-	assert.Equal(t, expiredCount, 2, "Expected 2 items to be expired")
+
+	lock.Lock()
+	defer lock.Unlock()
+	assert.Equal(t, 2, expiredCount, "Expected 2 items to be expired")
 }
 
 func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 	expiredCount := 0
+	var lock sync.Mutex
+
 	cache := NewCache()
 	cache.SetTTL(time.Duration(50 * time.Millisecond))
 	cache.SetCheckExpirationCallback(func(key string, value interface{}) bool {
@@ -148,12 +161,16 @@ func TestCacheCheckExpirationCallbackFunction(t *testing.T) {
 		return false
 	})
 	cache.SetExpirationCallback(func(key string, value interface{}) {
+		lock.Lock()
 		expiredCount = expiredCount + 1
+		lock.Unlock()
 	})
 	cache.Set("key", "value")
 	cache.Set("key2", "value")
 	<-time.After(110 * time.Millisecond)
-	assert.Equal(t, expiredCount, 1, "Expected 1 item to be expired")
+	lock.Lock()
+	assert.Equal(t, 1, expiredCount, "Expected 1 item to be expired")
+	lock.Unlock()
 }
 
 func TestCacheNewItemCallbackFunction(t *testing.T) {
@@ -167,7 +184,7 @@ func TestCacheNewItemCallbackFunction(t *testing.T) {
 	cache.Set("key2", "value")
 	cache.Set("key", "value")
 	<-time.After(110 * time.Millisecond)
-	assert.Equal(t, newItemCount, 2, "Expected only 2 new items")
+	assert.Equal(t, 2, newItemCount, "Expected only 2 new items")
 }
 
 func TestCacheRemove(t *testing.T) {
@@ -178,8 +195,8 @@ func TestCacheRemove(t *testing.T) {
 	<-time.After(70 * time.Millisecond)
 	removeKey := cache.Remove("key")
 	removeKey2 := cache.Remove("key_2")
-	assert.Equal(t, removeKey, true, "Expected 'key' to be removed from cache")
-	assert.Equal(t, removeKey2, false, "Expected 'key_2' to already be expired from cache")
+	assert.Equal(t, true, removeKey, "Expected 'key' to be removed from cache")
+	assert.Equal(t, false, removeKey2, "Expected 'key_2' to already be expired from cache")
 }
 
 func TestCacheSetWithTTLExistItem(t *testing.T) {
@@ -189,8 +206,8 @@ func TestCacheSetWithTTLExistItem(t *testing.T) {
 	<-time.After(30 * time.Millisecond)
 	cache.SetWithTTL("key", "value2", time.Duration(50*time.Millisecond))
 	data, exists := cache.Get("key")
-	assert.Equal(t, exists, true, "Expected 'key' to exist")
-	assert.Equal(t, data.(string), "value2", "Expected 'data' to have value 'value2'")
+	assert.Equal(t, true, exists, "Expected 'key' to exist")
+	assert.Equal(t, "value2", data.(string), "Expected 'data' to have value 'value2'")
 }
 
 func BenchmarkCacheSetWithoutTTL(b *testing.B) {

--- a/item.go
+++ b/item.go
@@ -16,7 +16,9 @@ func newItem(key string, data interface{}, ttl time.Duration) *item {
 		ttl:  ttl,
 		key:  key,
 	}
+	item.mutex.Lock()
 	item.touch()
+	item.mutex.Unlock()
 	return item
 }
 
@@ -31,11 +33,9 @@ type item struct {
 
 // Reset the item expiration time
 func (item *item) touch() {
-	item.mutex.Lock()
 	if item.ttl > 0 {
 		item.expireAt = time.Now().Add(item.ttl)
 	}
-	item.mutex.Unlock()
 }
 
 // Verify if the item is expired

--- a/item.go
+++ b/item.go
@@ -1,7 +1,6 @@
 package ttlcache
 
 import (
-	"sync"
 	"time"
 )
 
@@ -16,9 +15,8 @@ func newItem(key string, data interface{}, ttl time.Duration) *item {
 		ttl:  ttl,
 		key:  key,
 	}
-	item.mutex.Lock()
+	// since nobody is aware yet of this item, it's safe to touch without lock here
 	item.touch()
-	item.mutex.Unlock()
 	return item
 }
 
@@ -27,7 +25,6 @@ type item struct {
 	data       interface{}
 	ttl        time.Duration
 	expireAt   time.Time
-	mutex      sync.Mutex
 	queueIndex int
 }
 
@@ -40,12 +37,9 @@ func (item *item) touch() {
 
 // Verify if the item is expired
 func (item *item) expired() bool {
-	item.mutex.Lock()
 	if item.ttl <= 0 {
-		item.mutex.Unlock()
 		return false
 	}
 	expired := item.expireAt.Before(time.Now())
-	item.mutex.Unlock()
 	return expired
 }

--- a/priority_queue.go
+++ b/priority_queue.go
@@ -3,6 +3,7 @@ package ttlcache
 import (
 	"container/heap"
 	"sync"
+	"time"
 )
 
 func newPriorityQueue() *priorityQueue {
@@ -42,8 +43,16 @@ func (pq priorityQueue) Len() int {
 	return length
 }
 
+// Less will consider items with time.Time default value (epoch start) as more than set items.
 func (pq priorityQueue) Less(i, j int) bool {
 	pq.mutex.Lock()
+	var empty time.Time
+	if pq.items[i].expireAt == empty {
+		return false
+	}
+	if pq.items[j].expireAt == empty {
+		return true
+	}
 	less := pq.items[i].expireAt.Before(pq.items[j].expireAt)
 	pq.mutex.Unlock()
 	return less

--- a/priority_queue.go
+++ b/priority_queue.go
@@ -2,7 +2,6 @@ package ttlcache
 
 import (
 	"container/heap"
-	"sync"
 	"time"
 )
 
@@ -13,7 +12,6 @@ func newPriorityQueue() *priorityQueue {
 }
 
 type priorityQueue struct {
-	mutex sync.Mutex
 	items []*item
 }
 
@@ -37,15 +35,12 @@ func (pq *priorityQueue) remove(item *item) {
 }
 
 func (pq priorityQueue) Len() int {
-	pq.mutex.Lock()
 	length := len(pq.items)
-	pq.mutex.Unlock()
 	return length
 }
 
 // Less will consider items with time.Time default value (epoch start) as more than set items.
 func (pq priorityQueue) Less(i, j int) bool {
-	pq.mutex.Lock()
 	var empty time.Time
 	if pq.items[i].expireAt == empty {
 		return false
@@ -54,33 +49,26 @@ func (pq priorityQueue) Less(i, j int) bool {
 		return true
 	}
 	less := pq.items[i].expireAt.Before(pq.items[j].expireAt)
-	pq.mutex.Unlock()
 	return less
 }
 
 func (pq priorityQueue) Swap(i, j int) {
-	pq.mutex.Lock()
 	pq.items[i], pq.items[j] = pq.items[j], pq.items[i]
 	pq.items[i].queueIndex = i
 	pq.items[j].queueIndex = j
-	pq.mutex.Unlock()
 }
 
 func (pq *priorityQueue) Push(x interface{}) {
-	pq.mutex.Lock()
 	item := x.(*item)
 	item.queueIndex = len(pq.items)
 	pq.items = append(pq.items, item)
-	pq.mutex.Unlock()
 }
 
 func (pq *priorityQueue) Pop() interface{} {
-	pq.mutex.Lock()
 	old := pq.items
 	n := len(old)
 	item := old[n-1]
 	item.queueIndex = -1
 	pq.items = old[0 : n-1]
-	pq.mutex.Unlock()
 	return item
 }

